### PR TITLE
[LoongArch] Support LoongArch-specific amswap[_db].{b/h} and amadd[_db].{b/h} instructions

### PR DESCRIPTION
**Pull request by tangaac**
Two options for clang: -mlam-bh & -mno-lam-bh.
Enable or disable amswap[__db].{b/h} and amadd[__db].{b/h} instructions.
The default is -mno-lam-bh.
Only works on LoongArch64.

### CODE REVIEW

File: llvm/lib/Target/LoongArch/LoongArchInstrInfo.td [hunk 1]
```diff
@@ -2088,8 +2088,26 @@ multiclass binary_atomic_op_wd<string inst, string op, string signed = ""> {
...
+
+  def : Pat<(!cast<PatFrag>(op#"_i8") GPR:$rj, GPR:$rk),
+            (!cast<Instruction>(inst#"__DB_B") GPR:$rk, GPR:$rj)>;
+  def : Pat<(!cast<PatFrag>(op#"_i16") GPR:$rj, GPR:$rk),
+            (!cast<Instruction>(inst#"__DB_H") GPR:$rk, GPR:$rj)>;
+}
+
 let Predicates = [IsLA64] in {
 
+let Predicates = [ HasLAM_BH ] in {
```
[the remark this answers is not in the document]
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArchInstrInfo.td]: It does.
I don't know this before,  it's updated now.

**Review by SixWeining — APPROVED** [review]

**Comment by cjappl** [comment]
Just a heads up, I think this broke documentation builds. After bisecting it points to this commit.

```
ninja -C build docs-clang-html
FAILED: tools/clang/docs/CMakeFiles/docs-clang-html ~/code/radsan_cjappl/llvm-project/build/tools/clang/docs/CMakeFiles/docs-clang-html
cd ~/code/radsan_cjappl/llvm-project/build/tools/clang/docs && /opt/homebrew/Cellar/cmake/3.28.3/bin/cmake -E env ~/.virtualenvs/llvm/bin/sphinx-build -b htm
l -d ~/code/radsan_cjappl/llvm-project/build/tools/clang/docs/_doctrees-clang-html -q -t builder-html -D version=20 -D release=20.0.0git -tPreRelease -W ~co
de/radsan_cjappl/llvm-project/build/tools/clang/docs ~code/radsan_cjappl/llvm-project/build/tools/clang/docs/html

Warning, treated as error:
~/code/radsan_cjappl/llvm-project/build/tools/clang/docs/ClangCommandLineReference.rst:4819:Unknown target name: "amswap". [docutils]
```

See an example here:
https://github.com/llvm/llvm-project/actions/runs/11502212789/job/32016658426?pr=112727

**Comment by SixWeining** [comment]
> Just a heads up, I think this broke documentation builds. After bisecting it points to this commit.

Thanks. Fixed by #113632.

**Comment by cjappl** [comment]
Thanks for the quick fix, working for me after I pulled latest main.

**Pull request merged**